dips/dip-5.md

```diff
@@ -0,0 +1,80 @@
+
+For a good couple of months, V2 sustained all the needs of P2P messaging on the DeSo blockchain. However, as more apps were built, particularly mobile applications, there was an increasing need to handle messages without having access to user's seed. One important use case were derived keys. Derived keys allowed applications to sign transactions on behalf of a user without accessing their main seed. But derived keys didn't give developers the ability to encrypt/decrypt messages, which were still being encrypted to the user's main public key. We certainly can't give applications access to user's main seed, and the current messaging protocol didn't give any leeway to circumvent these limitations, hence a new protocol altogether was needed.
+
+DeSo V3 messages solve the above problem, and give derived key holders the ability to handle user messages. We do this by introducing messaging keys, which will be registered on-chain and can be safely shared with third-party app developers. Messaging keys will be labeled and derived deterministically from user's main seed. If a user has registered a messaging key, messages can then be encrypted to this messaging key. Specifically, there will be a default messaging key, labeled "default-key", that the DeSo Identity Service will look for when handling messages. DeSo V3 Messages rollout is intended to be a non-forking change, so messaging keys will be passed in transaction's ExtraData, as opposed to in a new transaction.
+
+## Specification
+The change will involve all components of the DeSo stack, including Core, Backend, Identity, and Frontend. In addition, app developers should adjust how they communicate with the DeSo Identity Service to include the messaging party in message decryption. Here are Github pull requests for reference:
+- [DeSo core](https://github.com/deso-protocol/core/pull/170)
+- [DeSo backend](https://github.com/deso-protocol/backend/pull/253)
+- [DeSo identity](https://github.com/deso-protocol/identity/pull/107)
+- [DeSo node frontend](https://github.com/deso-protocol/frontend/pull/543)
+
+### Technical changes
+
+#### Core
+A conditional statement will be added in `_connectBasicTransfer()` to look for transaction's `ExtraData` and check for three fields: `MessagingPublicKey`, `MessagingKeyName`, and `MessagingKeySignature`. These respectively will contain the public key and its name along with an owner-signed hash of `(MessagingPublicKey + MessagingKeyName)`. The signature is provided because we want derived key holders to submit messaging keys; therefore, the transaction signature alone wouldn't be sufficient to prove authorization of the messaging key. Similarly, disconnecting a messaging key will happen in `_disconnectBasicTransfer()` and will delete the messaging entry from the `UtxoView` and DB. Core will only accept one key for a label, an attempt to submit a different public key for the same key name will silently error.
```

Comment:
I don't like the precedent this sets. Avoiding a fork by shoving information into a basic transfer's ExtraData map does not seem advisable. I would much prefer we just add a new transaction type and schedule a network upgrade.

Reply:
Also, if nodes are going to build these indexes correctly, we need them to upgrade accordingly so we may as well set a block height and schedule this change for the next fork.